### PR TITLE
hubble: support drop_reason_desc in flow filter

### LINKED ISSUE #32051: CFP: Support drop_reason_desc in flow filter
## Cilium Feature Proposal

**Is your proposed feature related to a problem?**

We are using [hubble exporter](https://docs.cilium.io/en/latest/observability/hubble-exporter/) and want to exclude events with `drop_reason_desc` `UNSUPPORTED_L3_PROTOCOL`.

We tried:

```yaml
hubble:
  export:
  # ...
      dynamic:
        enabled: true
        config:
         # ...
          content:
          - name: principal
            # ...
            excludeFilters: 
            - drop_reason_desc:
              - UNSUPPORTED_L3_PROTOCOL
            - drop_reason:
              - "139" # UNSUPPORTED_L3_PROTOCOL
            - event_type:
              - type: 1
                match_sub_type: true
                sub_type: 139
```

Only third (event_type) way works, but this is not user-friendly. The two others fail to load.

**Describe the feature you'd like**

`drop_reason_desc` to be supported as filter type.

See: https://github.com/cilium/cilium/blob/aef6814f7e895d644234c679a956ef86c832a1b8/api/v1/flow/flow.pb.go#L2892-L2980

### PR DESCRIPTION
**Pull request by chaunceyjiang**
Fixes: #32051

```
hubble:
  export:
  # ...
      dynamic:
        enabled: true
        config:
         # ...
          content:
          - name: principal
            # ...
            excludeFilters: 
            - drop_reason_desc:
              - UNSUPPORTED_L3_PROTOCOL
              - POLICY_DENY

```

```release-note
hubble: support drop\_reason\_desc in flow filter
```

### CODE REVIEW

File: api/v1/flow/flow.proto [hunk 1]
```diff
@@ -517,6 +517,8 @@ message FlowFilter {
 
     // only return Flows that were classified with a particular verdict.
     repeated Verdict verdict = 5;
+    // only applicable to Verdict = DROPPED.
+    repeated DropReason drop_reason_desc = 35;
```
Comment (chaunceyjiang) [inline api/v1/flow/flow.proto:521]: I'm not entirely sure whether the field name should be `drop_reason_desc` or `drop_reason` would be more appropriate here.

**Comment by chaunceyjiang** [comment]
@sathieu Hey, do you have a moment to give this PR a quick test?

File: api/v1/flow/flow.proto (hunk 1 above)
Comment (chancez) [reply api/v1/flow/flow.proto:521]: I think this is fine.

**Review by chancez — COMMENTED** [review]
I think this overall looks good, however I'm not really sure it's much of an improvement over filtering on `event_type`. I think the original request in 32051 might have been asking to support filtering on the stringified reason. 

Normally, in Hubble CLI we handle this for users, but since the original request is talking about filtering in the file export config, it's making me think we should be thinking about how to improve the UX.

Second, Hubble CLI has been merged into cilium/cilium, so we should probably add support for the new filter in the Hubble CLI too.

**Comment by chaunceyjiang** [comment]
> I think the original request in 32051 might have been asking to support filtering on the stringified reason.

Yes, so I'm not quite sure whether `repeated DropReason drop_reason_desc = 35;` or `repeated string drop_reason_desc = 35;` is a more appropriate solution.

**Comment by chaunceyjiang** [comment]
> Second, Hubble CLI has been merged into cilium/cilium, so we should probably add support for the new filter in the Hubble CLI too.

Ok,  If this PR is successfully merged, I'll submit another PR to complete this task.

**Comment by chaunceyjiang** [comment]
>  I think the original request in 32051 might have been asking to support filtering on the stringified reason.
Normally, in Hubble CLI we handle this for users, but since the original request is talking about filtering in the file export config, it's making me think we should be thinking about how to improve the UX.

Yeah, so I'm not sure which option is better - `repeated DropReason drop_reason_desc = 35;` or `repeated string drop_reason_desc = 35;`. Or maybe we can support both?

**Comment by michi-covalent** [comment]
thanks for picking this up @chaunceyjiang!

yes please make the cli-side change in https://github.com/cilium/cilium/tree/main/hubble in this pull request to make it easier to try this out.

in terms of the ux, the user should be able to specify the `"drop_reason_desc"` field returned in the json output. so for example, if the json output is:
```
  "drop_reason_desc": "POLICY_DENIED",
```
the CLI filter should be
```
hubble observe -t drop --drop-reason-desc POLICY_DENIED
```
the same for the export filter:
```
- drop_reason_desc:
  - POLICY_DENIED
```

and finally please include some sample usage / output in the pull request description to make it easier for other people to try this out 🚀🙏

**Comment by gandro** [comment]
> Yeah, so I'm not sure which option is better - `repeated DropReason drop_reason_desc = 35;` or `repeated string drop_reason_desc = 35;`. Or maybe we can support both?

As for string vs. DropReason - unless I'm mistaken, the JSON/YAML decoder should correctly convert the string to a `DropReason` in Golang/Protobuf. In other words, with `repeated DropReason drop_reason_desc` I expect the example `excludeFilters` from #32051 to "just work".

So I highly prefer to use a strongly typed value for the field.

**Comment by chancez** [comment]
> > Yeah, so I'm not sure which option is better - `repeated DropReason drop_reason_desc = 35;` or `repeated string drop_reason_desc = 35;`. Or maybe we can support both?
> 
> As for string vs. DropReason - unless I'm mistaken, the JSON/YAML decoder should correctly convert the string to a `DropReason` in Golang/Protobuf. In other words, with `repeated DropReason drop_reason_desc` I expect the example `excludeFilters` from #32051 to "just work".
> 
> So I highly prefer to use a strongly typed value for the field.

That's super good to know. For reference: https://github.com/protocolbuffers/protobuf-go/blob/v1.33.0/encoding/protojson/decode.go#L491-L517 we can see the protobuf unmarshaller handles enums as strings and numbers, using the enum name for strings, and enum integer for numbers.

**Review by chancez — APPROVED** [review]

**Review by gandro — APPROVED** [review]

**Comment by michi-covalent** [comment]
hey @chaunceyjiang you need to run `/test` to run all the required workflows.

**Comment by chaunceyjiang** [comment]
Okay, but I'm currently facing an issue where I've noticed that a few lines of code in the `hubble/cmd/observe_help.txt` file seem to have been accidentally changed.

The `hubble/cmd/observe_help.txt` file is one that I generated, but in my local environment, only L43 has been modified. However, I'm not sure why L147 also appears to have changed, even though I can't see any modified characters.

@michi-covalent 
```
make hubble 
./hubble observe -h > cmd/observe_help.txt
```

**Comment by michi-covalent** [comment]
ah ok. are you running this on a windows machine? i see some diff running from a mac:
```
% ./hubble observe -h > cmd/observe_help.txt
% git diff
diff --git a/hubble/cmd/observe_help.txt b/hubble/cmd/observe_help.txt
index 5220435dad..bc85ebf0d8 100644
--- a/hubble/cmd/observe_help.txt
+++ b/hubble/cmd/observe_help.txt
@@ -142,15 +142,15 @@ Server Flags:
       --request-timeout duration      Unary Request timeout. Only applies to non-streaming RPCs (ServerStatus, ListNodes, ListNamespaces). (default 12s)
       --server string                 Address of a Hubble server. Ignored when --input-file is provided. (default "localhost:4245")
       --timeout duration              Hubble server dialing timeout (default 5s)
-      --tls                           Specify that TLS must be used when establishing a connection to a Hubble server.
+      --tls                           Specify that TLS must be used when establishing a connection to a Hubble server.^M
                                       By default, TLS is only enabled if the server address starts with 'tls://'.
-      --tls-allow-insecure            Allows the client to skip verifying the server's certificate chain and host name.
-                                      This option is NOT recommended as, in this mode, TLS is susceptible to machine-in-the-middle attacks.
+      --tls-allow-insecure            Allows the client to skip verifying the server's certificate chain and host name.^M
+                                      This option is NOT recommended as, in this mode, TLS is susceptible to machine-in-the-middle attacks.^M
                                       See also the 'tls-server-name' option which allows setting the server name.
       --tls-ca-cert-files strings     Paths to custom Certificate Authority (CA) certificate files.The files must contain PEM encoded data.
-      --tls-client-cert-file string   Path to the public key file for the client certificate to connect to a Hubble server (implies TLS).
+      --tls-client-cert-file string   Path to the public key file for the client certificate to connect to a Hubble server (implies TLS).^M
                                       The file must contain PEM encoded data.
-      --tls-client-key-file string    Path to the private key file for the client certificate to connect a Hubble server (implies TLS).
+      --tls-client-key-file string    Path to the private key file for the client certificate to connect a Hubble server (implies TLS).^M
                                       The file must contain PEM encoded data.
       --tls-server-name string        Specify a server name to verify the hostname on the returned certificate (eg: 'instance.hubble-relay.cilium.io').
```
if it's easier i can amend the commit.

**Comment by chaunceyjiang** [comment]
@michi-covalent Thank you so much, I really appreciate the reminder. I've already taken care of that issue!

File: api/v1/flow/flow.proto (hunk 1 above)
Comment (kaworu) [reply api/v1/flow/flow.proto:521]: Why 35 though? As far as I can see the highest proto field number for `FlowFilter` before this patch was `http_header = 32`.

Comment (chaunceyjiang) [reply api/v1/flow/flow.proto:521]: ` repeated string cel_expression = 33;` 

CEL has used 33, considering that CEL is an experimental feature just introduced, therefore 34 is reserved for the future expansion of CEL.

Comment (kaworu) [reply api/v1/flow/flow.proto:521]: `cel_expression` is part of the `Experimental` message and thus doesn't have to follow the `FlowFilter` protobuf field numbers, added a commit in [this PR](https://github.com/cilium/cilium/pull/32130) to fix both https://github.com/cilium/cilium/pull/32130/commits/6a8f572d47f9aa417fbab9fee2dd3b13aa0819d6.

**Pull request merged**